File: python/open_guardrail/guards/context_window_utilization.py
"""Detect inefficient context window usage or overflow."""
from __future__ import annotations

import re
import time
from typing import List

from open_guardrail.core import GuardResult

_CTX_SECTIONS: list[re.Pattern[str]] = [
    re.compile(
        r"(?:^|\n)Context:\s*([\s\S]+?)(?:\n\n|$)",
        re.IGNORECASE,
    ),
    re.compile(
        r"(?:^|\n)Retrieved:\s*([\s\S]+?)(?:\n\n|$)",
        re.IGNORECASE,
    ),
    re.compile(
        r"(?:^|\n)Documents?:\s*([\s\S]+?)(?:\n\n|$)",
        re.IGNORECASE,
    ),
    re.compile(
        r"(?:^|\n)Passages?:\s*([\s\S]+?)(?:\n\n|$)",
        re.IGNORECASE,
    ),
    re.compile(
        r"(?:^|\n)Sources?:\s*([\s\S]+?)(?:\n\n|$)",
        re.IGNORECASE,
    ),
]


def _count_words(text: str) -> int:
    return len(text.split())


def _estimate_tokens(text: str) -> int:
    import math

    return math.ceil(_count_words(text) * 1.3)


def _detect_repetition(text: str) -> int:
    sentences = [
        s
        for s in re.split(r"[.!?]\s+", text)
        if len(s) > 20
    ]
    if len(sentences) < 4:
        return 0
    seen: set[str] = set()
    dupes = 0
    for s in sentences:
        norm = s.lower().strip()
        if norm in seen:
            dupes += 1
        else:
            seen.add(norm)
    return dupes


class _ContextWindowUtilization:
    def __init__(
        self,
        *,
        action: str = "warn",
        max_context_ratio: float = 0.8,
        estimated_window_size: int = 128000,
    ) -> None:
        self.name = "context-window-utilization"
        self.action = action
        self._max_ratio = max_context_ratio
        self._window = estimated_window_size

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: List[str] = []

        est_tokens = _estimate_tokens(text)
        utilization = est_tokens / self._window

        if utilization > self._max_ratio:
            pct = round(utilization * 100)
            issues.append(
                f"Near context overflow: ~{est_tokens}"
                f" tokens ({pct}% of {self._window})"
            )

        dupes = _detect_repetition(text)
        if dupes > 3:
            issues.append(
                f"Repeated content: {dupes}"
                " duplicate sentences"
            )

        ctx_words = 0
        for pat in _CTX_SECTIONS:
            m = pat.search(text)
            if m:
                ctx_words += _count_words(m.group(1))

        total_words = _count_words(text)
        if ctx_words > 0 and total_words > 0:
            resp_words = total_words - ctx_words
            if ctx_words > 500 and resp_words < 20:
                issues.append(
                    f"Poor utilization: {ctx_words}"
                    f" context words but only"
                    f" {resp_words} response words"
                )

        triggered = len(issues) > 0
        score = (
            min(utilization, 1.0) if triggered else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="context-window-utilization",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Context window utilization issue"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "issues": issues,
                    "estimated_tokens": est_tokens,
                    "utilization": round(utilization, 2),
                    "window_size": self._window,
                }
                if triggered
                else None
            ),
        )


def context_window_utilization(
    *,
    action: str = "warn",
    max_context_ratio: float = 0.8,
    estimated_window_size: int = 128000,
) -> _ContextWindowUtilization:
    return _ContextWindowUtilization(
        action=action,
        max_context_ratio=max_context_ratio,
        estimated_window_size=estimated_window_size,
    )

File: packages/guards/src/context-window-utilization.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ContextWindowUtilizationOptions {
  action: 'block' | 'warn';
  maxContextRatio?: number;
  estimatedWindowSize?: number;
}

const CONTEXT_SECTION_PATTERNS: RegExp[] = [
  /(?:^|\n)Context:\s*([\s\S]+?)(?:\n\n|$)/i,
  /(?:^|\n)Retrieved:\s*([\s\S]+?)(?:\n\n|$)/i,
  /(?:^|\n)Documents?:\s*([\s\S]+?)(?:\n\n|$)/i,
  /(?:^|\n)Passages?:\s*([\s\S]+?)(?:\n\n|$)/i,
  /(?:^|\n)Sources?:\s*([\s\S]+?)(?:\n\n|$)/i,
];

function countWords(text: string): number {
  return text.split(/\s+/).filter(Boolean).length;
}

function estimateTokens(text: string): number {
  return Math.ceil(countWords(text) * 1.3);
}

function detectRepetition(text: string): number {
  const sentences = text.split(/[.!?]\s+/).filter((s) => s.length > 20);
  if (sentences.length < 4) return 0;

  const seen = new Set<string>();
  let dupes = 0;
  for (const s of sentences) {
    const norm = s.toLowerCase().trim();
    if (seen.has(norm)) dupes++;
    else seen.add(norm);
  }
  return dupes;
}

export function contextWindowUtilization(options: ContextWindowUtilizationOptions): Guard {
  const maxRatio = options.maxContextRatio ?? 0.8;
  const windowSize = options.estimatedWindowSize ?? 128000;

  return {
    name: 'context-window-utilization',
    version: '0.1.0',
    description: 'Detects inefficient context window usage or near-overflow',
    category: 'content',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      const estTokens = estimateTokens(text);
      const utilization = estTokens / windowSize;

      if (utilization > maxRatio) {
        issues.push(
          `Near context overflow: ~${estTokens} tokens (${Math.round(utilization * 100)}% of ${windowSize})`,
        );
      }

      const dupes = detectRepetition(text);
      if (dupes > 3) {
        issues.push(`Repeated content detected: ${dupes} duplicate sentences`);
      }

      let contextWords = 0;
      for (const pat of CONTEXT_SECTION_PATTERNS) {
        const m = pat.exec(text);
        if (m) {
          contextWords += countWords(m[1]);
        }
      }

      const totalWords = countWords(text);
      if (contextWords > 0 && totalWords > 0) {
        const responseWords = totalWords - contextWords;
        if (contextWords > 500 && responseWords < 20) {
          issues.push(
            `Poor utilization: ${contextWords} context words but only ${responseWords} response words`,
          );
        }
      }

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(utilization, 1.0) : 0;

      return {
        guardName: 'context-window-utilization',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? {
              issues,
              estimatedTokens: estTokens,
              utilization: Math.round(utilization * 100) / 100,
              windowSize,
            }
          : undefined,
      };
    },
  };
}
